### PR TITLE
Use FrozenDictionary

### PR DESCRIPTION
**Pull request by MikeAlhayek**
@sebastienros It would be nice to see if the signification performance gain when switching to Immutable collections when possible.

I am guessing we'll only see real difference if we have lots of parts registered

### CODE REVIEW

**Comment by dannyd89** [comment]
How is this a performance gain? `ImmutableDictionary` is slower than a normal `Dictionary`, for example see the discussion [here](https://github.com/dotnet/runtime/issues/47812). If you want faster Read times and we dont care about the time it takes to build the dictionary, why not use a `FrozenDictionary`?

Sorry for being curious

**Comment by sebastienros** [comment]
This PR is creating the collections/dictionaries once so this is fine. Better use `FrozenDictionary` however since they are more optimized for fast read-only operations (lookups).

**Comment by dannyd89** [comment]
Yeah the benchmark really paints the picture, thanks! 
`ImmutableDictionary` and in this regard anything starting with `Immutable...` are really niche. Most of the time it's just better to use `Dictionary` or now `FrozenDictionary`.

Are there currently some works to improve the performance of OC in general? Any tracking issues where I could also contribute?

**Comment by MikeAlhayek** [comment]
@dannyd89 there isn't really a list of performance based issues for OC. We like to merge PR #14572 as it is expected to bring notable improvement since serialization and deserialization is something we heavily use in OC and YesSql.

If you want to improve OC, the todo list is quite lengthy. you can filter by `1.x` milestone on the issues and cherry pick which every issue you want to work on. We love any contributions that would improve OC https://github.com/OrchardCMS/OrchardCore/issues?q=is%3Aopen+is%3Aissue+milestone%3A1.x

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Services/LuceneIndexManager.cs [hunk 1]
```diff
@@ -192,7 +192,7 @@ public void Read(string indexName, Action<IndexReader> reader)
         /// </summary>
         public IReadOnlyDictionary<string, DateTime> GetTimestamps()
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Services/LuceneIndexManager.cs:193]: Delete this method, it is not used.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search.Lucene/Services/LuceneIndexManager.cs]: @sebastienros done. Are you good with merging this PR?

**Review by sebastienros — APPROVED** [review]

**Comment by sebastienros** [comment]
NB: You see how the `SortBuilders` property is used during configuration and then revoked once the public properties are used (and transformed to FrozenDictionary). I think this is the same pattern that should be used for JsonSerialization types, instead of registering these in DI, just have the collection initialized in Options. And the DI will resolve an options object instead of a collection of services (cheaper).

**Pull request merged**